examples/cis3-nft-sponsored-txs/src/lib.rs

```diff
@@ -0,0 +1,2083 @@
+    builder: &mut StateBuilder<S>,
+    logger: &mut impl HasLogger,
+) -> ContractResult<()> {
+    // Update the operator in the state.
+    match update {
+        OperatorUpdate::Add => state.add_operator(&sender, &operator, builder),
+        OperatorUpdate::Remove => state.remove_operator(&sender, &operator),
+    }
+
+    // Log the appropriate event
+    logger.log(&Cis2Event::<ContractTokenId, ContractTokenAmount>::UpdateOperator(
+        UpdateOperatorEvent {
+            owner: sender,
+            operator,
+            update,
+        },
+    ))?;
+
+    Ok(())
+}
+
+/// Enable or disable addresses as operators of the sender address.
+/// Logs an `UpdateOperator` event.
+///
+/// It rejects if:
+/// - It fails to parse the parameter.
+/// - Fails to log event.
+#[receive(
+    contract = "cis3_nft",
+    name = "updateOperator",
+    parameter = "UpdateOperatorParams",
+    error = "ContractError",
+    enable_logger,
+    mutable
+)]
+fn contract_update_operator<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+) -> ContractResult<()> {
+    // Parse the parameter.
+    let UpdateOperatorParams(params) = ctx.parameter_cursor().get()?;
+    // Get the sender who invoked this contract function.
+    let sender = ctx.sender();
+    let (state, builder) = host.state_and_builder();
+    for param in params {
+        _update_operator(param.update, sender, param.operator, state, builder, logger)?;
+    }
+    Ok(())
+}
+
+/// The parameter type for the contract function `permitUpdateOperator`.
+#[derive(Debug, Serialize, SchemaType)]
+pub struct PermitUpdateOperatorParams(
+    #[concordium(size_length = 2)] pub Vec<PermitUpdateOperatorParam>,
+);
+
+/// Verify an ed25519 signature and enable or disable addresses as operators of
+/// the given account. Execute a list of updates, in the order of the list.
+///
+/// Logs an `UpdateOperator` event.
+///
+/// It rejects if:
+/// - It fails to parse the parameter.
+/// - A different nonce is expected.
+/// - The signature was intended for a different contract.
+/// - The signature was intended for a different `entry_point`.
+/// - No public key was registered for the given account.
+/// - The signature can not be validated.
+/// - Fails to log event.
+#[receive(
+    contract = "cis3_nft",
+    name = "permitUpdateOperator",
+    parameter = "PermitUpdateOperatorParams",
+    crypto_primitives,
+    mutable,
+    enable_logger
+)]
+fn contract_permit_update_operator<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> ContractResult<()> {
+    // Parse the parameter.
+    let PermitUpdateOperatorParams(params) = ctx.parameter_cursor().get()?;
+
+    for param in params {
+        // Get the next consecutive nonce.
+        let nonce = host
```

Comment:
If `params.len > 1` then it is cheaper to keep the `nonce` in a variable outside the loop and only write it to the state once the for loop has finished.

Reply:
The `param.owner` can be different in every loop, so also the nonce can be different in every loop and that is why the nonce has to be inside of the loop and updated for every iteration in the for loop.

Reply:
Ah, right. That was a blunder on my part ;)